Solaris and error: 'NAN' was not declared in this scope

Hi Everyone,

I'm updating OpenSSH on Solaris 11.4 i86. libexpat is a dependency. libexpat 2.4.5 is having a little trouble:

```
$ make check -k
...
/bin/g++ -DHAVE_CONFIG_H -I. -I..   -DXML_ENABLE_VISIBILITY=1 -I./../lib -I/opt/ssh/incl
ude -DNDEBUG -Wall   -fexceptions -fno-strict-aliasing -g2 -O2 -m64 -march=native -fPIC
-pthread -MT runtestspp.o -MD -MP -MF .deps/runtestspp.Tpo -c -o runtestspp.o runtestspp
.cpp
In file included from runtestspp.cpp:36:0:
runtests.c: In function 'void test_billion_laughs_attack_protection_api()':
runtests.c:11712:32: error: 'NAN' was not declared in this scope
          parserWithoutParent, NAN)
                               ^
runtests.c:11734:32: error: 'INFINITY' was not declared in this scope
          parserWithoutParent, INFINITY)
                               ^
*** Error code 1
```

This patch appears to clear the issue. The self tests build and run without problems:

```diff
$ cat expat.patch 
--- tests/runtests.c
+++ tests/runtests.c 
@@ -60,6 +60,16 @@
 #  include <stdbool.h>
 #endif
 
+#if defined(__sun__)
+#  include <sunmath.h>
+#  if ! defined(INFINITY)
+#    define INFINITY (__builtin_huge_val())
+#  endif
+#  if ! defined(NAN)
+#    define NAN (__builtin_nan(""))
+#  endif
+#endif
+
 #include "expat.h"
 #include "chardata.h"
 #include "structdata.h"
@@ -6636,7 +6646,8 @@
   XML_Char **elemstr = (XML_Char **)userData;
   if (xcstrcmp(elemstr[0], name) != 0) {
     char buffer[1024];
-    sprintf(buffer, "unexpected end string: '%" XML_FMT_STR "'", name);
+    snprintf(buffer, sizeof(buffer), "unexpected end string: '%" XML_FMT_STR "'", name);
+    buffer[sizeof(buffer)-1] = '\0';  /* ensure NULL termination */
     fail(buffer);
   }
   triplet_end_flag = XML_TRUE;
```

@hartwork,

This is what I see with libexpat 2.4.5:

```
$ make check 2>&1 | fold -w 80
Making check in lib
Making check in examples
Making check in tests
Making check in .
make  runtests runtestspp
`runtests' is up to date.
/bin/g++ -DHAVE_CONFIG_H -I. -I..   -DXML_ENABLE_VISIBILITY=1 -I./../lib -I/expo
rt/home/jwalton/tmp/ok2delete/include -DNDEBUG -Wall   -fexceptions -fno-strict-
aliasing -g2 -O2 -m64 -march=native -fPIC -pthread -MT runtestspp.o -MD -MP -MF 
.deps/runtestspp.Tpo -c -o runtestspp.o runtestspp.cpp
In file included from runtests.c:49:0,
                 from runtestspp.cpp:36:
runtests.c: In function 'float portableINFINITY()':
runtests.c:11699:32: error: 'isinf' was not declared in this scope
   assert(isinf(handmadeINFINITY));
                                ^
```

> #include <ieeefp.h>
>
> int isinf(double x) { return !finite(x) && x==x; }

Ah, that just kicks the can down the road:

```
runtests.c: In function 'int isinf(double)':
runtests.c:11689:46: error: 'finite' was not declared in this scope
 static int isinf(double x) { return !finite(x) && x==x; }
                                              ^
```

@hartwork,

Yes, the last patched compiled and successfully completed its self tests. I did not know that's the portable way to get infinity and nan.

There were some minor warnings during build:

```
/bin/gcc -DHAVE_CONFIG_H -I. -I..  -DXML_ENABLE_VISIBILITY=1 -I./../lib -I/export/home/jwalton/tmp/ok2delete/include -DNDEBUG -Wall   -fexceptions -fno-strict-aliasing -Wmissing-prototypes -Wstrict-prototypes -pedantic -Wlogical-op -Wjump-misses-init -Wdouble-promotion -Wshadow -Wformat=2 -Wno-pedantic-ms-format -fvisibility=hidden -g2 -O2 -m64 -march=native -fPIC -pthread -std=c99 -MT runtests.o -MD -MP -MF .deps/runtests.Tpo -c -o runtests.o runtests.c
runtests.c:11689:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 portableNAN() {
 ^
runtests.c:11694:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
 portableINFINITY() {
 ^
cc1: warning: unrecognized command line option "-Wno-pedantic-ms-format" [enabled by default]
```

> Yes, the last patched compiled and successfully completed its self tests.

@noloader cool, thank you!

> I did not know that's the portable way to get infinity and nan.

It's one way that felt worth trying.

> That seems like something that should be in Gnulib. Do you use it?

Only once. But's let focus on Expat for now.

> There were some minor warnings during build:
> 
> ```
> /bin/gcc -DHAVE_CONFIG_H -I. -I..  -DXML_ENABLE_VISIBILITY=1 -I./../lib -I/export/home/jwalton/tmp/ok2delete/include -DNDEBUG -Wall   -fexceptions -fno-strict-aliasing -Wmissing-prototypes -Wstrict-prototypes -pedantic -Wlogical-op -Wjump-misses-init -Wdouble-promotion -Wshadow -Wformat=2 -Wno-pedantic-ms-format -fvisibility=hidden -g2 -O2 -m64 -march=native -fPIC -pthread -std=c99 -MT runtests.o -MD -MP -MF .deps/runtests.Tpo -c -o runtests.o runtests.c
> runtests.c:11689:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
>  portableNAN() {
>  ^
> runtests.c:11694:1: warning: function declaration isn't a prototype [-Wstrict-prototypes]
>  portableINFINITY() {
>  ^
> cc1: warning: unrecognized command line option "-Wno-pedantic-ms-format" [enabled by default]
> ```

That could be about missing `void`in the braces or so, and it may be legacy by now. I'll only address that if the CI says it too. We have `-Werror` in there.

--- a/expat/Changes
+++ b/expat/Changes
@@ -2,6 +2,11 @@ NOTE: We are looking for help with a few things:
       https://github.com/libexpat/libexpat/labels/help%20wanted
       If you can help, please get in touch.  Thanks!
 
+Release x.x.x xxx xxxxxxxx xx xxxx
+        Other changes:
+       #569 #571  tests: Resolve use of macros NAN and INFINITY for GNU G++
+                    4.8.2 on Solaris.
+
 Release 2.4.6 Sun February 20 2022
         Bug fixes:
             #566  Fix a regression introduced by the fix for CVE-2022-25313

--- a/expat/tests/runtests.c
+++ b/expat/tests/runtests.c
@@ -54,7 +54,6 @@
 #include <ctype.h>
 #include <limits.h>
 #include <stdint.h> /* intptr_t uint64_t */
-#include <math.h>   /* NAN, INFINITY, isnan */
 
 #if ! defined(__cplusplus)
 #  include <stdbool.h>
@@ -11762,6 +11761,16 @@ START_TEST(test_accounting_precision) {
 }
 END_TEST
 
+static float
+portableNAN() {
+  return strtof("nan", NULL);
+}
+
+static float
+portableINFINITY() {
+  return strtof("infinity", NULL);
+}
+
 START_TEST(test_billion_laughs_attack_protection_api) {
   XML_Parser parserWithoutParent = XML_ParserCreate(NULL);
   XML_Parser parserWithParent
@@ -11780,7 +11789,7 @@ START_TEST(test_billion_laughs_attack_protection_api) {
       == XML_TRUE)
     fail("Call with non-root parser is NOT supposed to succeed");
   if (XML_SetBillionLaughsAttackProtectionMaximumAmplification(
-          parserWithoutParent, NAN)
+          parserWithoutParent, portableNAN())
       == XML_TRUE)
     fail("Call with NaN limit is NOT supposed to succeed");
   if (XML_SetBillionLaughsAttackProtectionMaximumAmplification(
@@ -11802,7 +11811,7 @@ START_TEST(test_billion_laughs_attack_protection_api) {
       == XML_FALSE)
     fail("Call with positive limit >=1.0 is supposed to succeed");
   if (XML_SetBillionLaughsAttackProtectionMaximumAmplification(
-          parserWithoutParent, INFINITY)
+          parserWithoutParent, portableINFINITY())
       == XML_FALSE)
     fail("Call with positive limit >=1.0 is supposed to succeed");
 

